File: model_1/icpc-jakarta-2018/ice_1_solution.cpp
#include <iostream>
#include <vector>
#include <cstring>
using namespace std;

const int MAX = 505;

bool dfs(int u, bool seen[], vector<int> graph[], int matchR[]) {
    for (int v : graph[u]) {
        if (!seen[v]) {
            seen[v] = true;
            if (matchR[v] == -1 || dfs(matchR[v], seen, graph, matchR)) {
                matchR[v] = u;
                return true;
            }
        }
    }
    return false;
}

int main() {
    int R, C;
    cin >> R >> C;
    char grid[MAX][MAX];
    for (int i = 0; i < R; i++) {
        cin >> grid[i];
    }

    bool rowEmpty[MAX], colEmpty[MAX];
    memset(rowEmpty, true, sizeof(rowEmpty));
    memset(colEmpty, true, sizeof(colEmpty));

    for (int i = 0; i < R; i++) {
        for (int j = 0; j < C; j++) {
            if (grid[i][j] == '#') {
                rowEmpty[i] = false;
                colEmpty[j] = false;
            }
        }
    }

    vector<int> emptyRows;
    vector<int> emptyCols;
    for (int i = 0; i < R; i++) {
        if (rowEmpty[i]) {
            emptyRows.push_back(i);
        }
    }
    for (int j = 0; j < C; j++) {
        if (colEmpty[j]) {
            emptyCols.push_back(j);
        }
    }

    int A = emptyRows.size();
    int B = emptyCols.size();

    vector<int> graph[MAX];
    for (int i = 0; i < A; i++) {
        int r = emptyRows[i];
        for (int j = 0; j < B; j++) {
            int c = emptyCols[j];
            if (grid[r][c] == '.') {
                graph[i].push_back(j);
            }
        }
    }

    int matchR[MAX];
    memset(matchR, -1, sizeof(matchR));
    int matchingCount = 0;
    bool seen[MAX];

    for (int u = 0; u < A; u++) {
        memset(seen, 0, sizeof(seen));
        if (dfs(u, seen, graph, matchR)) {
            matchingCount++;
        }
    }

    int answer = A + B - matchingCount;
    cout << answer << endl;

    return 0;
}

File: model_1/icpc-jakarta-2018/ice_1_solution.py
import sys
from collections import deque

def main():
    data = sys.stdin.read().splitlines()
    if not data:
        return
    R, C = map(int, data[0].split())
    grid = []
    for i in range(1, 1+R):
        grid.append(data[i].strip())
    
    rowEmpty = [True] * R
    colEmpty = [True] * C

    for i in range(R):
        for j in range(C):
            if grid[i][j] == '#':
                rowEmpty[i] = False
                colEmpty[j] = False

    emptyRows = [i for i in range(R) if rowEmpty[i]]
    emptyCols = [j for j in range(C) if colEmpty[j]]
    
    A = len(emptyRows)
    B = len(emptyCols)
    
    if A == 0 and B == 0:
        print(0)
        return
        
    graph0 = [[] for _ in range(A)]
    for i in range(A):
        r = emptyRows[i]
        for j in range(B):
            c = emptyCols[j]
            if grid[r][c] == '.':
                graph0[i].append(j)
    
    if A == 0 or B == 0:
        print(A + B)
        return

    matchL = [-1] * A
    matchR = [-1] * B
    dist = [-1] * A
    matching_count = 0

    def bfs():
        nonlocal dist
        queue = deque()
        for i in range(A):
            if matchL[i] == -1:
                dist[i] = 0
                queue.append(i)
            else:
                dist[i] = -1
        found = False
        while queue:
            u = queue.popleft()
            for v in graph0[u]:
                u_next = matchR[v]
                if u_next == -1:
                    found = True
                elif u_next != -1 and dist[u_next] == -1:
                    dist[u_next] = dist[u] + 1
                    queue.append(u_next)
        return found

    def dfs(u):
        for v in graph0[u]:
            u_next = matchR[v]
            if u_next == -1:
                matchR[v] = u
                matchL[u] = v
                return True
            if dist[u_next] == dist[u] + 1:
                if dfs(u_next):
                    matchR[v] = u
                    matchL[u] = v
                    return True
        return False

    while bfs():
        for u in range(A):
            if matchL[u] == -1:
                if dfs(u):
                    matching_count += 1
                    
    ans = A + B - matching_count
    print(ans)

if __name__ == "__main__":
    main()